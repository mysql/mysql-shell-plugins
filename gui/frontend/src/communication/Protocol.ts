/*
 * Copyright (c) 2020, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

/* eslint-disable max-len */
/* eslint-disable @typescript-eslint/naming-convention */

import { ListenerEntry } from "../supplement/Dispatch";
import { uuid } from "../utilities/helpers";

export type ShellDictionaryType = string | number | boolean | undefined | unknown | null | IShellDictionary;
export interface IShellDictionary {
    [key: string]: ShellDictionaryType | ShellDictionaryType[];
}

export enum ShellPromptResponseType {
    Ok = "OK",
    Cancel = "CANCEL"
}

// The request structure sent to the backend. That's why it uses snake case.
export interface IShellRequest extends IShellDictionary {
    request_id: string; // A unique ID to identify this request. It's used for all responses.
    request: string;    // The requested operation.

    command?: string;        // Optional field to carry the command if this is an execution request.
    args?: IShellDictionary; // Optional arguments for the command.
}

//  Begin auto generated types

//  End auto generated types

export class Protocol {

    public static getRequestCommandExecute(command: string, rest: IShellDictionary): IShellRequest {
        return Protocol.getStandardRequest("execute", { command, ...rest });
    }

    public static getRequestUsersAuthenticate(username: string, password: string): IShellRequest {
        return Protocol.getStandardRequest("authenticate", {
            username,
            password,
        });
    }

    public static getRequestPromptReply(request_id: string, type: ShellPromptResponseType, reply: string, moduleSessionId: string): IShellRequest {
        const result = Protocol.getStandardRequest("prompt_reply", {
            request_id,
            type,
            reply,
            module_session_id: moduleSessionId,
        });

        return result;
    }

    public static getStandardRequest(request: string, rest?: IShellDictionary): IShellRequest {
        return {
            request_id: uuid(),
            request,
            ...rest,
        };
    }
}

export interface IPromptReplyBackend {
    sendReply: (requestId: string, type: ShellPromptResponseType, reply: string) => ListenerEntry;
}
