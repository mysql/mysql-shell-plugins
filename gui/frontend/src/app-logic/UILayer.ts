/*
 * Copyright (c) 2024, 2025, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import type { IStatusBarItem, StatusBarAlignment } from "../components/ui/Statusbar/StatusBarItem.js";
import type { IServicePasswordRequest } from "./general-types.js";

/**
 * The UI layer object comprises functionality that is specific to the environment it is used in.
 * This is only relevant for code which is shared between different platforms (e.g. the extension and the app).
 * The UI layer redirects requests for UI specific functionality to the appropriate implementation and must be set
 * up by the application logic layer (either the app host or the extension host).
 */
export interface IUILayer {
    /**
     * Shows an information notification. The message is automatically hidden after the given timeout
     * (in milliseconds, default is 5000).
     */
    showInformationNotification(message: string, timeout?: number): Promise<string | undefined>;

    /** Shows a warning notification. */
    showWarningNotification(message: string): Promise<string | undefined>;

    /** Shows a error notification. */
    showErrorNotification(message: string): Promise<string | undefined>;

    /**
     * Creates a status bar {@link StatusBarItem item}.
     *
     * @param alignment The alignment of the item.
     * @param priority The priority of the item. Higher values mean the item should be shown more to the left.
     * @returns A new status bar item.
     */
    createStatusBarItem(alignment?: StatusBarAlignment, priority?: number): IStatusBarItem;

    /**
     * Creates a status bar {@link StatusBarItem item}.
     *
     * @param id The unique identifier of the item.
     * @param alignment The alignment of the item.
     * @param priority The priority of the item. Higher values mean the item should be shown more to the left.
     * @returns A new status bar item.
     */
    createStatusBarItem(id: string, alignment?: StatusBarAlignment, priority?: number): IStatusBarItem;

    /**
     * Shows a message in the status bar.
     *
     * @param message The message to show.
     * @param timeout The time in milliseconds after which the message is automatically hidden. If not set, the message
     *               is shown until the next message is set.
     */
    setStatusBarMessage(message: string, timeout?: number): void;

    /**
     * Asks the user for a yes/no decision. Returns the yes/no string, depending on what the user clicked or
     * undefined if the user clicked outside/cancelled the dialog (e.g. pressed <escape>).
     */
    confirm(message: string, yes: string, no: string): Promise<string | undefined>;

    /**
     * Triggers a password request.
     *
     * @param request The request with details about the password prompt.
     *
     * @returns The entered password or undefined if the password dialog was cancelled.
     */
    requestPassword(request: IServicePasswordRequest): Promise<string | undefined>;
}

const defaultHandler = (): number => {
    console.error("Internal error: the UI layer has not been set up.");

    return -1;
};

const asyncDefaultHandler = (): Promise<string | undefined> => {
    console.error("Internal error: the UI layer has not been set up.");

    return Promise.resolve(undefined);
};

export let ui: IUILayer = {
    showInformationNotification: asyncDefaultHandler,
    showWarningNotification: asyncDefaultHandler,
    showErrorNotification: asyncDefaultHandler,
    createStatusBarItem: (): IStatusBarItem => {
        console.error("Internal error: the UI layer has not been set up.");

        return {} as IStatusBarItem;
    },
    setStatusBarMessage: defaultHandler,
    confirm: asyncDefaultHandler,
    requestPassword: asyncDefaultHandler,
};

export const registerUiLayer = (newLayer: IUILayer): void => {
    ui = newLayer;
};
