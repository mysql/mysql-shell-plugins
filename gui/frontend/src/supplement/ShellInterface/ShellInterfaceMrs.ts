/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { ListenerEntry } from "../Dispatch";
import { IShellDictionary, MessageScheduler, ProtocolMrs } from "../../communication";
import { webSession } from "../WebSession";

export class ShellInterfaceMrs {

    // The key under which the module session is stored in the WebSession instance.
    public moduleSessionLookupId = "";

    public listServices(): ListenerEntry {
        const request = ProtocolMrs.getRequestListServices({ moduleSessionId: this.moduleSessionId });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsListServices" });
    }

    public addService(urlContextRoot: string, urlProtocol: string, urlHostName: string, isDefault?: boolean,
        comments?: string, enabled?: boolean): ListenerEntry {
        const request = ProtocolMrs.getRequestAddService(urlContextRoot, urlHostName, enabled, {
            moduleSessionId: this.moduleSessionId,
            urlProtocol,
            isDefault,
            comments,
        });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsAddService" });
    }

    public updateServiceById(serviceId: number, value: IShellDictionary): ListenerEntry {
        const request = ProtocolMrs.getRequestUpdateService({
            moduleSessionId: this.moduleSessionId,
            serviceId,
            value: JSON.stringify({
                url_host: value.hostName,
                url_context_root: value.urlContextRoot,
                url_protocol: value.protocols,
                is_default: value.isDefault,
                enabled: value.enabled,
                comments: value.comments,
            }),
            urlContextRoot: undefined,
            urlHostName: undefined,
            interactive: false,
            raiseExceptions: false,
        });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsUpdateService" });
    }

    public deleteService(serviceId?: number): ListenerEntry {
        const request = ProtocolMrs.getRequestDeleteService({ moduleSessionId: this.moduleSessionId, serviceId });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsDeleteService" });
    }

    public setDefaultService(serviceId: number): ListenerEntry {
        const request = ProtocolMrs.getRequestSetServiceDefault({
            moduleSessionId: this.moduleSessionId,
            serviceId,
        });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsSetDefaultService" });
    }

    public listSchemas(serviceId: number): ListenerEntry {
        const request = ProtocolMrs.getRequestListSchemas({ serviceId, moduleSessionId: this.moduleSessionId });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsListSchemas" });
    }

    public deleteSchema(schemaId: number, serviceId: number): ListenerEntry {
        const request = ProtocolMrs.getRequestDeleteSchema({
            moduleSessionId: this.moduleSessionId,
            schemaId,
            serviceId,
        });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsDeleteSchema" });
    }

    public addSchema(schemaName: string, requestPath: string,
        requiresAuth: boolean, serviceId?: number, itemsPerPage?: number, comments?: string): ListenerEntry {
        const request = ProtocolMrs.getRequestAddSchema({
            moduleSessionId: this.moduleSessionId,
            schemaName,
            requestPath,
            requiresAuth,
            serviceId,
            itemsPerPage,
            comments,
        });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsAddSchema" });
    }

    public updateSchemaById(schemaId: number, value: IShellDictionary): ListenerEntry {

        const request = ProtocolMrs.getRequestUpdateSchema({
            moduleSessionId: this.moduleSessionId,
            schemaId,
            value: JSON.stringify({
                name: value.name,
                request_path: value.requestPath,
                requires_auth: value.requiresAuth,
                enabled: value.enabled,
                items_per_page: value.itemsPerPage,
                comments: value.comments,
            }),
            interactive: false,
            raiseExceptions: false,
        });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsUpdateSchema" });
    }

    public addDbObject(dbObjectName: string, dbObjectType: string, schemaName: string, autoAddSchema: boolean,
        requestPath: string, crudOperations: string[], crudOperationFormat: string, requiresAuth: boolean,
        rowUserOwnershipEnforced: boolean, rowUserOwnershipColumn?: string,
        schemaId?: number, itemsPerPage?: number, comments?: string): ListenerEntry {
        const request = ProtocolMrs.getRequestAddDbObject({
            moduleSessionId: this.moduleSessionId,
            dbObjectName,
            dbObjectType,
            schemaId,
            schemaName,
            autoAddSchema,
            requestPath,
            crudOperations,
            crudOperationFormat,
            requiresAuth,
            itemsPerPage,
            rowUserOwnershipEnforced,
            rowUserOwnershipColumn,
            comments,
        });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsAddService" });
    }

    public listDbObjects(schemaId: number): ListenerEntry {
        const request = ProtocolMrs.getRequestListDbObjects(schemaId, { moduleSessionId: this.moduleSessionId });

        return MessageScheduler.get.sendRequest(request, { messageClass: "mrsListDbObjects" });
    }

    private get moduleSessionId(): string | undefined {
        return webSession.moduleSessionId(this.moduleSessionLookupId);
    }
}
