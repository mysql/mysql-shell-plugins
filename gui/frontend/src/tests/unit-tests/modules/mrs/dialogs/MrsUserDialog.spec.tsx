/*
 * Copyright (c) 2024, 2025, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is designed to work with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have either included with
 * the program or referenced in the documentation.
 *
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { fireEvent, render } from "@testing-library/preact";
import { createRef } from "preact";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";

import { userEvent } from "@testing-library/user-event";
import { registerUiLayer } from "../../../../../app-logic/UILayer.js";
import { IShellDictionary } from "../../../../../communication/Protocol.js";
import {
    IMrsAuthAppData, type IMrsAuthVendorData, type IMrsContentFileData,
    type IMrsContentSetData, type IMrsDbObjectData, type IMrsRoleData, type IMrsRouterData,
    type IMrsRouterService, type IMrsSchemaData, type IMrsServiceData, type IMrsStatusData,
    type IMrsTableColumnWithReference, type IMrsUserData
} from "../../../../../communication/ProtocolMrs.js";
import { MrsHub } from "../../../../../modules/mrs/MrsHub.js";
import { ShellInterfaceMrs } from "../../../../../supplement/ShellInterface/ShellInterfaceMrs.js";
import { ShellInterfaceSqlEditor } from "../../../../../supplement/ShellInterface/ShellInterfaceSqlEditor.js";
import { uiLayerMock } from "../../../__mocks__/UILayerMock.js";
import {
    authAppsData, mrsContentFileData, mrsContentSetData, mrsDbObjectData, mrsRouterData, mrsSchemaData,
    mrsServiceData, mrsServicesData, mrsStatusMock, mrsUserData, routerServiceData
} from "../../../data-models/data-model-test-data.js";
import { DialogHelper, mockClassMethods, nextRunLoop } from "../../../test-helpers.js";

mockClassMethods(ShellInterfaceMrs, {
    status: (): Promise<IMrsStatusData> => {
        return Promise.resolve(mrsStatusMock);
    },
    listServices: (): Promise<IMrsServiceData[]> => {
        return Promise.resolve(mrsServicesData);
    },
    listSchemas: (): Promise<IMrsSchemaData[]> => {
        return Promise.resolve(mrsSchemaData);
    },
    listRouters: (): Promise<IMrsRouterData[]> => {
        return Promise.resolve(mrsRouterData);
    },
    listContentSets: (): Promise<IMrsContentSetData[]> => {
        return Promise.resolve(mrsContentSetData);
    },
    listUsers: (): Promise<IMrsUserData[]> => {
        return Promise.resolve(mrsUserData);
    },
    listContentFiles: (): Promise<IMrsContentFileData[]> => {
        return Promise.resolve(mrsContentFileData);
    },
    getService: (): Promise<IMrsServiceData> => {
        return Promise.resolve(mrsServiceData);
    },
    getSchema: (): Promise<IMrsSchemaData> => {
        return Promise.resolve(mrsSchemaData[0]);
    },
    getRouterServices: (): Promise<IMrsRouterService[]> => {
        return Promise.resolve(routerServiceData);
    },
    listAuthApps: (): Promise<IMrsAuthAppData[]> => {
        return Promise.resolve(authAppsData);
    },
    listAppServices: (_appId?: string): Promise<IMrsServiceData[]> => {
        return Promise.resolve([]);
    },
    listRoles: (): Promise<IMrsRoleData[]> => {
        return Promise.resolve([{
            id: "role1",
            derivedFromRoleId: "",
            specificToServiceId: "",
            caption: "unknown role",
            description: "",
        }]);
    },
    listDbObjects: (schemaId: string): Promise<IMrsDbObjectData[]> => {
        return Promise.resolve(mrsDbObjectData);
    },
    updateAuthApp: (): Promise<void> => {
        return Promise.resolve();
    },
    getAuthVendors: (): Promise<IMrsAuthVendorData[]> => {
        return Promise.resolve([{
            id: "MRS",
            name: "MRS",
            enabled: true,
        }]);
    },
    getTableColumnsWithReferences: (): Promise<IMrsTableColumnWithReference[]> => {
        return Promise.resolve([{
            position: 1,
            name: "actorId",
            refColumnNames: "actor_id",
            tableSchema: "myschema",
            tableName: "actor",
        }]);
    },
    dumpSdkServiceFiles: vi.fn().mockResolvedValue(true),
    addService: (urlContextRoot: string, name: string,
        urlProtocol: string[], urlHostName: string,
        comments: string, enabled: boolean, options: IShellDictionary | null,
        authPath: string, authCompletedUrl: string,
        authCompletedUrlValidation: string, authCompletedPageContent: string,
        metadata?: IShellDictionary, published = false): Promise<IMrsServiceData> => {
        return Promise.resolve({
            enabled: Number(enabled),
            published: Number(published),
            hostCtx: "",
            id: "",
            isCurrent: 1,
            urlContextRoot,
            urlHostName,
            urlProtocol: urlProtocol.join(","),
            comments,
            options: options ?? {},
            authPath,
            authCompletedUrl,
            authCompletedUrlValidation,
            authCompletedPageContent,
            enableSqlEndpoint: 0,
            customMetadataSchema: "",
            metadata,
            name,
        });
    },
    setCurrentService: (serviceId: string): Promise<void> => {
        return Promise.resolve();
    },
    linkAuthAppToService: (appId: string, serviceId: string): Promise<void> => {
        return Promise.resolve();
    },
    addUser: () => {
        return Promise.resolve();
    },
});

describe("MRS User dialog tests", () => {
    const localEvent = userEvent.setup();

    const hubRef = createRef<MrsHub>();
    let dialogHelper: DialogHelper;
    const backend = new ShellInterfaceSqlEditor();

    let unmount: () => boolean;

    beforeAll(() => {
        registerUiLayer(uiLayerMock);

        const result = render(<MrsHub ref={hubRef} />);
        unmount = result.unmount;

        dialogHelper = new DialogHelper("mrsUserDialog", "MySQL REST User");
    });

    afterAll(() => {
        unmount();

        vi.resetAllMocks();
    });

    it("Show MRS User Dialog (snapshot) and escape", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);

        dialogHelper.actOnDialogAppearance("valueEditDialog", (portal) => {
            expect(portal).toMatchSnapshot();
            const cancelButton = portal.querySelector("#cancel");
            fireEvent.click(cancelButton!);
        });

        await hubRef.current!.showMrsUserDialog(backend, authAppsData[0]);

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
    });

    it("Dialog error testing", async () => {
        let portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);

        dialogHelper.actOnDialogAppearance("valueEditDialog", async (portal) => {
            const okButton = portal.querySelector("#ok");

            await localEvent.clear(portal.querySelector(".input#name")!);
            await localEvent.clear(portal.querySelector(".input#authString")!);
            fireEvent.click(okButton!);
            await nextRunLoop();
            dialogHelper.verifyErrors([
                "The authentication string is required for this app.",
                "The user name or email are required for this app.",
            ]);

            await dialogHelper.setInputText("name", "My User");
            fireEvent.click(okButton!);
            await nextRunLoop();
            dialogHelper.verifyErrors(["The authentication string is required for this app."]);

            await dialogHelper.setInputText("authString", "MySQLR0cks!");
            fireEvent.click(okButton!);
            await nextRunLoop();
            dialogHelper.verifyErrors();

            fireEvent.click(okButton!);
        });

        await hubRef.current!.showMrsUserDialog(backend, authAppsData[0]);

        portals = document.getElementsByClassName("portal");
        expect(portals).toHaveLength(0);
    });

});
