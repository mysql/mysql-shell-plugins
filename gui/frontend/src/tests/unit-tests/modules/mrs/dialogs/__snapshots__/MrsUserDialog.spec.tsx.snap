// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MRS User dialog tests Show MRS User Dialog (snapshot) and escape 1`] = `
<div
  class="msg portal"
  id="mrsUserDialog"
  style="--background-opacity: 0.5;"
>
  <div
    class="msg dialog visible valueEditDialog"
  >
    <div
      class="title verticalCenterContent"
    >
      <div
        class="msg icon codicon codicon-terminal"
      />
      <label
        class="msg label"
        data-tooltip="expand"
      >
        MySQL REST User
      </label>
      <div
        class="msg button imageOnly"
        id="closeButton"
        role="button"
        tabindex="0"
      >
        <div
          class="msg icon"
          style="mask-image: url("close2.svg");"
        />
      </div>
    </div>
    <div
      class="content fixedScrollbar"
    >
      <div
        class="msg grid"
        id="contentGrid"
        style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
      >
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: row; grid-column: span 8;"
        >
          <label
            class="msg label sectionTitle"
            data-tooltip="expand"
          >
            Enter new MySQL REST User Values
          </label>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 3;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            User Name
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <input
                class="msg input valueEditor"
                id="name"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The name of the user
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 3;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            User Password
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <input
                class="msg input valueEditor"
                id="authString"
                spellcheck="true"
                type="password"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The password of the user
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 2;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Authentication App
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <input
                class="msg input valueEditor"
                id="authApp"
                readonly=""
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The authentication app
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 3;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Email
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <input
                class="msg input valueEditor"
                id="email"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              The email of the user
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 3;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Assigned Roles
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar tagInput dropdown multiSelect"
              id="roles"
              style="flex-direction: row; flex-wrap: wrap;"
              tabindex="0"
            >
              <label
                class="msg label tag"
                data-tooltip="expand"
                draggable="true"
              >
                Full Access
              </label>
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              Roles assign DB Object privileges to the user.
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: center; align-items: stretch; grid-column: span 2;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Â 
          </label>
          <div
            class="control-wrap"
          >
            <label
              class="msg label description"
              data-tooltip="expand"
            />
            <label
              class="msg checkbox checked valueEditor"
              id="loginPermitted"
              tabindex="0"
            >
              <span
                class="checkMark"
              />
              Permit Login
            </label>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              Allow user to log in
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-row: span 2; grid-column: span 6;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            User Options
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <textarea
                class="msg input valueEditor"
                id="appOptions"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              Additional options in JSON format
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 2;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Vendor User Id
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <input
                class="msg input valueEditor"
                id="vendorUserId"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              Set by OAuth2 vendors
            </label>
          </div>
        </div>
        <div
          class="msg container fixedScrollbar gridCell"
          style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 2;"
        >
          <label
            class="msg label valueTitle"
            data-tooltip="expand"
          >
            Mapped User Id
          </label>
          <div
            class="control-wrap"
          >
            <div
              class="msg container fixedScrollbar inputWithProgress"
              style="flex-direction: row;"
            >
              <input
                class="msg input valueEditor"
                id="mappedUserId"
                spellcheck="true"
                type="text"
              />
            </div>
            <label
              class="msg label cellDescription"
              data-tooltip="expand"
            >
              Optional id for sync-ing
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="footer verticalCenterContent"
    >
      <div
        class="msg container fixedScrollbar leftItems"
        style="flex-direction: row;"
      />
      <div
        class="msg container fixedScrollbar rightItems"
        style="flex-direction: row-reverse;"
      >
        <div
          caption="OK"
          class="msg button"
          id="ok"
          role="button"
          tabindex="0"
        >
          OK
        </div>
        <div
          caption="Cancel"
          class="msg button"
          id="cancel"
          role="button"
          tabindex="0"
        >
          Cancel
        </div>
      </div>
    </div>
  </div>
</div>
`;
